### DISCUSSION TITLE
Mixing a One-To-Many relation  into an abstract base class

### DISCUSSION THREAD

**Discussion by radiantone** [sqlalchemy/sqlalchemy#7064] [Usage Questions] [answered]
Hi

Having trouble with this scenario.

I have an abstract base class for all my models. Defines things like id, owner, date.

I want to add a mixin class to this base class that adds a one to many relation to another model "Log".

Goal being each model class will now have a "logs" relation without me modifying each separately.

Problem arises because my base abstract class doesnt have a table so completing the relationship with the proper backref table and foreign key id is the problem.

Surely there is a way to do this?

I read all the docs but none had this scenario - mixing into an abstract class.

**Comment by zzzeek** [comment, marked as answer]
you use `@declared_attr` for this as described at https://docs.sqlalchemy.org/en/14/orm/declarative_mixins.html#mixing-in-relationships 

when you talk about a "backref" to an unmapped abstract class, that's not possible because this class doesnt represent anything persisted in the database.  you can only make backrefs on Log that are specific to each class as in the example below.

I suspect you are close to a set of patterns we call "polymorphic associations" or "generic associations" so you probably want to check out the examples at https://docs.sqlalchemy.org/en/14/orm/examples.html#module-examples.generic_associations

simple version of what you describe is as follows:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import ForeignKey
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base, Session
from sqlalchemy.orm import declared_attr
from sqlalchemy.orm import relationship

Base = declarative_base()


class Abstract(Base):
    __abstract__ = True

    id = Column(Integer, primary_key=True)

class HasLog:
    @declared_attr
    def log_id(cls):
        return Column(ForeignKey("log.id"))

    @declared_attr
    def log(cls):
        return relationship("Log", backref=f"logged_{cls.__name__}")


class Log(Base):
    __tablename__ = "log"

    id = Column(Integer, primary_key=True)


class A(HasLog, Abstract):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)

class B(HasLog, Abstract):
    __tablename__ = "b"

    id = Column(Integer, primary_key=True)
    data = Column(String)


e = create_engine("sqlite://", echo=True)
Base.metadata.create_all(e)

s = Session(e)

l1, l2 = Log(), Log()


a1 = A(log=l1)
a2 = A(log=l2)
b1 = B(log=l1)
b2 = B(log=l1)

s.add_all([
    a1, a2, b1, b2, l1, l2
])

s.commit()

assert set(l1.logged_A) == {a1}
assert set(l1.logged_B) == {b1, b2}
```

**Reply by radiantone** [reply]
Thank you so much for quick reply!! You rock!

**Reply by radiantone** [reply]
Hi again! I tried this, but the problem still is it needs to be a one-to-many relation where there is one class (e.g. A, B) to many Log. In your example, log_id gets added to any class with HasLog, but it needs to be a relationship "logs" that is a list of Log rows.

Problem I had was how to declare the back relationship in HasLog so that it maps back to the subclass and not the base class because the relationship wants to know what table the back id is in, but the abstract base class doesn't have a table. Also, my id field is ONLY in the Base class, not each individual subclass. This is why I was trying to do the mixin on the abstract base class. Thoughts?

**Reply by zzzeek** [reply]
> Hi again! I tried this, but the problem still is it needs to be a one-to-many relation where there is one class (e.g. A, B) to many Log. In your example, log_id gets added to any class with HasLog, but it needs to be a relationship "logs" that is a list of Log rows.

yup, that's generic association.  see the examples I linked previously.    table_per_related is IMO the best one for simple cases though people at first find it counter-intuitive.     the different versions are chosen based on how you need to deal with these Log entries, i.e. are you going to be querying Log objects across all parent types at once, or not.

> 
> Problem I had was how to declare the back relationship in HasLog so that it maps back to the subclass and not the base class because the relationship wants to know what table the back id is in, but the abstract base class doesn't have a table. Also, my id field is ONLY in the Base class, not each individual subclass. This is why I was trying to do the mixin on the abstract base class. Thoughts?

see the examples, it uses HasAddresses.  change that to HasLogs

**Reply by radiantone** [reply]
Hi! I followed the example and it worked! Two observations.
1. I have to add the mixin to all my concrete classes and not the abstract base class. This is "ok", but am hoping someday I can add it to the base class to avoid direct exposure to subclasses.
2. Is there any way to remove the subclass specific association table and use a column in the Address (Log for me) table that points back to the parent? This way I can have all my logs in one table and do easier query/reporting/aggregations on that data.

**Reply by zzzeek** [reply]
1. you can add the mixin to an abstract base class, shouldnt make any difference
2. there are two other versions of the recipe , discriminator_on_association and generic_fk, which as i mentioned allow you to query all Log entries at once.   "generic_fk" is relationally incorrect, by the way, we add it there since django and rails both feature this incorrect pattern.

**Reply by radiantone** [reply]
I was able to accomplish my goals using this

```
class HasLogs(object):

    @declared_attr
    def logs(cls):
        return relationship(
            "LogModel",
            order_by="desc(LogModel.created)",
            primaryjoin="foreign(LogModel.oid) == remote(%s.id)" % cls.__name__
        )
```
